File: scala-server/app/com/bolour/boardgame/scala/server/domain/Scorer.scala
package com.bolour.boardgame.scala.server.domain

import com.bolour.boardgame.scala.common.domain._
import com.bolour.boardgame.scala.common.domain.ScoreMultiplierType._
import com.bolour.boardgame.scala.common.domain.ScoreMultiplier._
import org.slf4j.LoggerFactory

/**
  * Scores plays.
  *
  * Note that a play is scored before its moved pieces are
  * laid down on the board.
  *
  * @param dimension Needed for determining cross words to be scored.
  * @param trayCapacity Need to add the bonus when all tiles are used in a play.
  */
class Scorer(val dimension: Int, trayCapacity: Int) {

  val logger = LoggerFactory.getLogger(this.getClass)

  val multiplierGrid: Grid[ScoreMultiplier] = mkMultiplierGrid(dimension)

  /**
    * Score a play.
    *
    * @param crossWordFinder Finds newly-created crosswords to be included in scoring.
    * @param playPieces Consecutive list of all play pieces for a play.
    *                   Includes both moved and existing pieces forming the play word.
    * @return The complete score of the play.
    */
  def scorePlay(crossWordFinder: CrossWordFinder, playPieces: List[PlayPiece]): Int = {
    val crossingPlays = crossWordFinder.findCrossPlays(playPieces)
    logger.debug(s"crossing plays: ${crossingPlays}")
    val crossScoreList = crossingPlays filter { cp => cp.length > 1 } map { cp => scoreWord(cp)}
    logger.debug(s"crossing score list: ${crossScoreList}")
    val crossWordsScore = crossScoreList.sum
    val wordScore = scoreWord(playPieces map { pp => (pp.piece.value, pp.point, pp.moved) })
    logger.debug(s"principle word score: ${wordScore}")
    wordScore + crossWordsScore
  }

  /**
    * Low-level method for scoring individual words formed in a play.
    *
    * @param playInfo List of data about the positions and movements of the word's letters.
    * @return The score of the word.
    */
  def scoreWord(playInfo: List[(Char, Point, Boolean)]): Int = {
    val multipliers = playInfo map {_._2} map { multiplierGrid.cell(_) }
    val playPieceMultipliers = playInfo zip multipliers
    val letterScores = playPieceMultipliers map {
      case ((letter, _, moved), multiplier) =>
        val factor = if (moved && multiplier.isLetterMultiplier) multiplier.factor else 1
        factor * Piece.worths(letter)
    }
    val baseScore = letterScores.sum

    val aggregateWordMultiplier = (playPieceMultipliers map {
      case ((_, _, moved), multiplier) =>
        if (moved && multiplier.isWordMultiplier) multiplier.factor else 0
    }).sum

    var score = Math.max(1, aggregateWordMultiplier) * baseScore
    val numMoves = playInfo.count(_._3)
    if (numMoves == trayCapacity)
      score += Scorer.Bonus
    score
  }
}

object Scorer {
  type Score = Int
  val Bonus = 50
  def apply(dimension: Int, trayCapacity: Int): Scorer = new Scorer(dimension, trayCapacity)
}

File: haskell-server/src/BoardGame/Server/Domain/Scorer.hs
--
-- Copyright 2017 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- TODO. Create tests for the scorer module.
-- Pay particular attention to representation of empty cells in the implementation.
-- Then connect this module to the rest of the application and use it to compute scores.
module BoardGame.Server.Domain.Scorer (
    Scorer
  , Scorer(scorePlay, scoreWord)
  , mkScorer
  ) where

import Data.Maybe (fromJust)
import BoardGame.Common.Domain.ScoreMultiplier (ScoreMultiplier, ScoreMultiplier(ScoreMultiplier), noMultiplier)
import qualified BoardGame.Common.Domain.ScoreMultiplier as ScoreMultiplier
import BoardGame.Common.Domain.Piece (Piece)
import qualified BoardGame.Common.Domain.Piece as Piece
import BoardGame.Common.Domain.Point (Point)
import BoardGame.Common.Domain.PlayPiece (PlayPiece, PlayPiece(PlayPiece), MoveInfo)
import qualified BoardGame.Common.Domain.PlayPiece as PlayPiece
import BoardGame.Common.Domain.Grid (Grid)
import BoardGame.Server.Domain.Board (Board)
import qualified BoardGame.Server.Domain.CrossWordFinder as CrossWordFinder
import qualified BoardGame.Common.Domain.Grid as Grid

bonus = 50

data Scorer = Scorer {
    scorePlay :: Board -> [PlayPiece] -> Int
  , scoreWord :: [(Char, Point, Bool)] -> Int
}

mkScorer :: Int -> Int -> Scorer
mkScorer dimension trayCapacity =
  let multGrid = ScoreMultiplier.mkMultiplierGrid dimension
  in Scorer
       (doScorePlay dimension trayCapacity multGrid)
       (doScoreWord dimension trayCapacity multGrid)

doScorePlay :: Int -> Int -> Grid ScoreMultiplier -> Board -> [PlayPiece] -> Int
doScorePlay dimension trayCapacity multGrid board playPieces =
  let crossPlays = CrossWordFinder.findCrossPlays board playPieces
      calcWordScore :: [MoveInfo] -> Int = doScoreWord dimension trayCapacity multGrid
      isCrossWordPlay :: [MoveInfo] -> Bool = (> 1) . length
      crossScores = calcWordScore <$> filter isCrossWordPlay crossPlays
      crossWordsScore = sum crossScores
      wordScore = calcWordScore $ PlayPiece.toMoveInfo <$> playPieces
  in wordScore + crossWordsScore

doScoreWord :: Int -> Int -> Grid ScoreMultiplier -> [MoveInfo] -> Int
doScoreWord dimension trayCapacity multGrid moves =
  let points = (\(_, point, _) -> point) <$> moves
      multipliers = Grid.cell multGrid <$> points
      moveMultipliers = moves `zip` multipliers -- List[((letter, point, moved), scoreMultiplier)]

      -- Get the base letter score.
      baseScore = scoreLetters moveMultipliers

      -- Multiply by the total word multiplier factors for new pieces.
      calcWordMultFactor ((_, _, moved), mult @ ScoreMultiplier {factor}) =
        if moved && ScoreMultiplier.isWordMultiplier mult then factor else 0
      sumWordFactors = max 1 (sum $ calcWordMultFactor <$> moveMultipliers)
      wordScore = sumWordFactors * baseScore

      -- Add the full tray play bonus.
      numMoves = length $ filter (\(_, _, moved) -> moved) moves
      bonusScore = if numMoves == trayCapacity then bonus else 0
      score = wordScore + bonusScore
  in score

scoreLetters :: [(MoveInfo, ScoreMultiplier)] -> Int
scoreLetters moveMultipliers =
  sum (letterScore <$> moveMultipliers)
    where
      letterScore ((letter, _, moved), mult @ ScoreMultiplier {factor}) =
        let w = Piece.letterWorth letter
            useFactor = ScoreMultiplier.isLetterMultiplier mult
        in w * if moved && useFactor then factor else 1
